cou/utils/juju_utils.py

```diff
@@ -509,3 +516,295 @@ def _check_time(self) -> None:
             raise TimeoutException(
                 f"Model {self.model_name} has not stabilized after {self.timeout} seconds."
             )
+
+
+def retry(
+    function: Optional[Callable] = None,
+    timeout: int = DEFAULT_TIMEOUT,
+    no_retry_exception: Optional[tuple] = None,
+) -> Callable:
+    """Retry function for usage in COUModel.
+
+    :param function: function to wrapped
+    :type function: Optional[Callable]
+    :param timeout: timeout in seconds
+    :type timeout: int
+    :param no_retry_exception: tuple of exception on which function will not be retried
+    :type no_retry_exception: Optional[tuple]
+    :return: wrapped function
+    :rtype: Callable
+    """
+    ignored_exceptions = (TimeoutException, *(no_retry_exception or ()))
```

Comment:
I did not understand why we raising timeout exception. Assume we lost our network connection. It will create TimeoutException and we will not be able to retry. We should only raise TimeoutException from line 556.

Reply:
The connection lost will not create `TimeoutException` maybe `TimeoutError`, the `TimeoutException` is our own exception raised only by this retry decorator. Why it's needed is due nested decorating. For example we are using function `_get_model` in multiple other functions and if we decide to decorate any of it (as I did for `set_application_config`) you do not want to retry if function `_get_model` failed with `TimeoutException`.